consensus/polybft/bridge_manager.go

```diff
@@ -0,0 +1,488 @@
+	// defaultMaxAttemptsToSend specifies how many sending retries for one transaction
+	defaultMaxAttemptsToSend = uint64(15)
+	// defaultMaxEventsPerBatch specifies maximum events per one batchExecute tx
+	defaultMaxEventsPerBatch = uint64(10)
+)
+
+var (
+	stateSyncEventSig           = new(contractsapi.StateSyncedEvent).Sig()
+	checkpointSubmittedEventSig = new(contractsapi.CheckpointSubmittedEvent).Sig()
+	exitProcessedEventSig       = new(contractsapi.ExitProcessedEvent).Sig()
+)
+
+// BridgeBackend is an interface that defines functions required by bridge components
+type BridgeBackend interface {
+	Runtime
+	StateSyncProofRetriever
+	ExitEventProofRetriever
+}
+
+// RelayerEventMetaData keeps information about a relayer event
+type RelayerEventMetaData struct {
+	EventID     uint64 `json:"eventID"`
+	CountTries  uint64 `json:"countTries"`
+	BlockNumber uint64 `json:"blockNumber"` // block when event is sent
+	SentStatus  bool   `json:"sentStatus"`
+}
+
+func (ed RelayerEventMetaData) String() string {
+	return fmt.Sprintf("%d", ed.EventID)
+}
+
+// relayerConfig is a struct that holds the relayer configuration
+type relayerConfig struct {
+	maxBlocksToWaitForResend uint64
+	maxAttemptsToSend        uint64
+	maxEventsPerBatch        uint64
+	eventExecutionAddr       types.Address
+}
+
+// eventTrackerConfig is a struct that holds the event tracker configuration
+type eventTrackerConfig struct {
+	consensus.EventTracker
+
+	jsonrpcAddr           string
+	startBlock            uint64
+	stateSenderAddr       types.Address
+	checkpointManagerAddr types.Address
+	exitHelperAddr        types.Address
+	trackerPollInterval   time.Duration
+}
+
+// RelayerState is an interface that defines functions that a relayer store has to implement
+type RelayerState interface {
+	GetAllAvailableRelayerEvents(limit int) (result []*RelayerEventMetaData, err error)
+	UpdateRelayerEvents(events []*RelayerEventMetaData, removeIDs []uint64, dbTx *bolt.Tx) error
+}
+
+// relayerEventsProcessor is a parent struct of both state sync and exit relayer
+// that holds functions common to both relayers
+type relayerEventsProcessor struct {
+	logger     hclog.Logger
+	state      RelayerState
+	blockchain blockchainBackend
+
+	config *relayerConfig
+	sendTx func([]*RelayerEventMetaData) error
+}
+
+// ProcessEvents processes all relayer events that were either successfully or unsuccessfully executed
+// and executes all the events that can be executed in regards to relayerConfig
+func (r *relayerEventsProcessor) processEvents() {
+	// we need twice as batch size because events from first batch are possible already sent maxAttemptsToSend times
+	events, err := r.state.GetAllAvailableRelayerEvents(int(r.config.maxEventsPerBatch) * 2)
+	if err != nil {
+		r.logger.Error("retrieving events failed", "err", err)
+
+		return
+	}
+
+	removedEventIDs := []uint64{}
+	sendingEvents := []*RelayerEventMetaData{}
+	currentBlockNumber := r.blockchain.CurrentHeader().Number
```

Comment:
Maybe provide the initial capacity to the worst-case scenario:
```suggestion
	removedEventIDs := make([]uint64, 0, len(events))
	sendingEvents := make([]*RelayerEventMetaData, 0, len(events))
```

Reply:
Fix: https://github.com/Ethernal-Tech/blade/pull/47/commits/64db03d69359faf64096b84b14e11291939e97d4